File: api/server/src/main/java/su/plo/voice/api/server/event/connection/TcpPacketSendEvent.java
package su.plo.voice.api.server.event.connection;

import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import su.plo.voice.api.event.EventCancellableBase;
import su.plo.voice.api.server.player.VoiceServerPlayer;
import su.plo.voice.proto.packets.Packet;

/**
 * This event is fired when the server is about to send a {@link Packet} to the player.
 */
@RequiredArgsConstructor
public final class TcpPacketSendEvent extends EventCancellableBase {

    @Getter
    private final @NonNull VoiceServerPlayer player;
    @Getter
    private final @NonNull Packet<?> packet;
}

File: api/proxy/src/main/kotlin/su/plo/voice/api/proxy/event/connection/TcpPacketSendEvent.kt
package su.plo.voice.api.proxy.event.connection

import su.plo.voice.api.event.EventCancellableBase
import su.plo.voice.api.server.player.VoicePlayer
import su.plo.voice.proto.packets.Packet

/**
 * This event is fired when the server is about to send a [Packet] to the player.
 *
 * This event is fired ONLY when invoked by [VoicePlayer.sendPacket].
 * If it's forwarded from the server (server -> proxy -> player) this event won't be fired.
 */
data class TcpPacketSendEvent(
    val player: VoicePlayer,
    val packet: Packet<*>
) : EventCancellableBase()